sqlglot/dialects/dialect.py

```diff
@@ -1073,36 +1073,47 @@ def property_sql(self: Generator, expression: exp.Property) -> str:
     return f"{self.property_name(expression, string_key=True)}={self.sql(expression, 'value')}"
 
 
-def str_position_sql(
+def strposition_sql(
     self: Generator,
     expression: exp.StrPosition,
-    generate_instance: bool = False,
-    str_position_func_name: str = "STRPOS",
+    func_name: str = "STRPOS",
+    supports_position: bool = False,
+    supports_occurrence: bool = False,
 ) -> str:
-    this = self.sql(expression, "this")
-    substr = self.sql(expression, "substr")
-    position = self.sql(expression, "position")
-    instance = expression.args.get("instance") if generate_instance else None
-    position_offset = ""
-
-    if position:
-        # Normalize third 'pos' argument into 'SUBSTR(..) + offset' across dialects
-        this = self.func("SUBSTR", this, position)
-        position_offset = f" + {position} - 1"
-
-    strpos_sql = self.func(str_position_func_name, this, substr, instance)
-
-    if position_offset:
-        zero = exp.Literal.number(0)
-        # If match is not found (returns 0) the position offset should not be applied
-        case = exp.If(
-            this=exp.EQ(this=strpos_sql, expression=zero),
+    string = expression.this
+    substr = expression.args.get("substr")
+    position = expression.args.get("position")
+    occurrence = expression.args.get("occurrence")
+    zero = exp.Literal.number(0)
+    one = exp.Literal.number(1)
+
+    if supports_occurrence and occurrence and supports_position and not position:
+        position = one
+
+    if position and not supports_position:
+        string = exp.Substring(this=string, start=position)
+
+    args = [substr, string] if func_name in ["LOCATE", "CHARINDEX"] else [string, substr]
+    if supports_position:
+        args.append(position)
+    if supports_occurrence:
+        args.append(occurrence)
```

Comment:
We can do `elif occurrence: self.unsupported(...)` in L1101, to ensure we won't silently drop an `occurrence` argument, if present but not supported.